src/main/java/org/opensearch/ad/rest/handler/AbstractAnomalyDetectorActionHandler.java

```diff
@@ -215,21 +217,42 @@ public AbstractAnomalyDetectorActionHandler(
 
     /**
      * Start function to process create/update/validate anomaly detector request.
-     * Check if anomaly detector index exist first, if not, will create first.
-     * 
-     * @throws IOException IOException from {@link AnomalyDetectionIndices#initAnomalyDetectorIndexIfAbsent(ActionListener)}
+     * If detector is not using custom result index, check if anomaly detector
+     * index exist first, if not, will create first. Otherwise, check if custom
+     * result index exists or not. If exists, will check if index mapping matches
+     * AD result index mapping and if user has correct permission to write index.
+     * If doesn't exist, will create custom result index with AD result index
+     * mapping.
      */
-    public void start() throws IOException {
-        if (!anomalyDetectionIndices.doesAnomalyDetectorIndexExist() && !this.isDryRun) {
-            logger.info("AnomalyDetector Indices do not exist");
-            anomalyDetectionIndices
-                .initAnomalyDetectorIndex(
-                    ActionListener.wrap(response -> onCreateMappingsResponse(response, false), exception -> listener.onFailure(exception))
-                );
-        } else {
-            logger.info("AnomalyDetector Indices do exist, calling prepareAnomalyDetectorIndexing");
-            logger.info("DryRun variable " + this.isDryRun);
-            validateDetectorName(this.isDryRun);
+    public void start() {
+        String resultIndex = anomalyDetector.getResultIndex();
+        // use default detector result index which is system index
+        if (resultIndex == null) {
+            createOrUpdateDetector();
+            return;
+        }
+
+        // user custom result index
+        anomalyDetectionIndices.initCustomResultIndexAndExecute(resultIndex, () -> createOrUpdateDetector(), listener);
+    }
+
+    private void createOrUpdateDetector() {
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
```

Comment:
What was the reason you moved stashContext() from IndexAnomalyDetectorTransportAction to here

Reply:
If not move to here, all clients which needs to call `IndexAnomalyDetectorTransportAction` need to stash context first, and need to catch exception from all clients. More reasonable to move to here as it's always required to access detector index.